internal/restorer/fileswriter_windows.go

```diff
@@ -0,0 +1,35 @@
+package restorer
+
+import "os"
+
+// OpenFile opens the file with truncate and write only options.
+// In case of windows, it first attempts to open an existing file,
+// and only if the file does not exist, it opens it with create option
+// in order to create the file. This is done, otherwise if the ads stream
+// is written first (in which case it automatically creates an empty main
+// file before writing the stream) and then when the main file is written
+// later, the ads stream can be overwritten.
+func (*filesWriter) OpenFile(createSize int64, path string) (*os.File, error) {
+	var flags int
+	var f *os.File
+	var err error
+	// TODO optimize this. When GenericAttribute change is merged, we can
```

Comment:
We don't need the GenericAttribute to know whether a file is an ADS or not; looking a the file name is enough.

This change is only correct, if the target directory is empty, otherwise the file can already exist and even have the wrong file size! So, I'm afraid we'll have to teach the `filerestorer.go` how to properly handle the ordering constraints for ADS. (I'm not sure so far what's the best approach for that)

Reply:
Actually, this TODO is in reference to the main file which has the ADS attached. Without this the main file is oblivious to any ADS attached to it, the ADS themselves just exist as independent nodes. 
And we want to know that the main file has ADS attached, because only in that case do we want to check for presence of the main file before using the O_CREATE option. Otherwise, we can directly use the O_CREATE option like the existing master code.

----------
Regarding the "This change is only correct, if the target directory is empty, otherwise the file can already exist and even have the wrong file size", I'm not sure I understand. There may be some confusion.

Let's say we have a test file - **D:\targetfile.txt**. It has an ADS attached - **D:\targetfile.txt:Stream1:$DATA**.
Now this file is backed up along with its ADS and these are stored as 2 separate files in the restic repo.

During restore, this method OpenFile will be called twice, and the calls can come any order depending on the content:

Once, for file name - D:\\targetfile.txt
Another time, for the ADS - D:\\targetfile.txt:Stream1:$DATA

(For now, consider that the target restore directory did not already exist.)

Whichever happens first will find - os.IsNotExist(err) is true and will then call the code below resulting in the creation of the main file D:\targetfile.txt.

```
flags = os.O_CREATE | os.O_TRUNC | os.O_WRONLY
f, err = os.OpenFile(path, flags, 0600)
```


**If it was called for D:\\targetfile.txt first**, 
then D:\\targetfile.txt will be created and the contents of the main file will be written to it. 
Then the call comes for D:\targetfile.txt:Stream1:$DATA at some point, it will find that the file for the stream does not exist.
It will use the os.O_CREATE option and it creates the stream only and writes the data to the stream without any issue.

**If it was called for D:\\targetfile.txt:Stream1:$DATA first,**
then it finds that the stream file does not exist and it uses the os.O_CREATE flag, and in that case, it creates an empty file D:\\targetfile.txt and starts writing the ADS at path - D:\\targetfile.txt:Stream1:$DATA.
Next when the call comes in for the main file targetfile.txt, it will find that os.IsNotExist(err) is false, because the main file was already created as a 0 size file and it will not use the os.O_CREATE option in this case. It will just continue to write the contents on the existing 0 size targetfile.txt.
Without this new check in place, when the second call comes in for targetfile.txt, it would replace the 0 size file which was written when the write for  D:\\targetfile.txt:Stream1:$DATA had started and ultimately the ADS is lost during restore.

**Even if the target restore directory and all the files already existed before restore**, the first time the file is opened, createSize >= 0 and we initially open the file with **os.O_TRUNC** | os.O_WRONLY. 
So, the file is truncated before the new content for restore is written. The file sizes after restore are correct. 
Am I missing some scenario?

Reply:
There is another scenario that depends on the GenericAttribute.
If a file exists in target with 10 ADStreams and it needs to be replaced with an older file with 5 ADStreams during restore, we need to know when restoring the main file, that the extra 5 streams need to be removed.
For this the main file needs to know what ADS streams are attached to it. Remaining streams will be removed during restore of main file.

So, the main file will have a GenericAttribute - TypeADSMain with value as -
Stream1|Stream2|Stream3 ...
I will update the code once https://github.com/restic/restic/pull/4611 is merged.